### PR TITLE
Text Generation with Transformer Decoder example

### PR DESCRIPTION
**Pull request by jessechancy**
The colab I've written before to test out transformer decoder. I've formatted it to be one of the example on the keras.io website.

### CODE REVIEW

File: examples/nlp/text_generation_transformer_decoder.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        num_heads=NUM_HEADS,
+        intermediate_dim=FEED_FORWARD_DIM,
+    )
+    x = transformer_block(x) # no cross-attention
+  # output
+  outputs = keras.layers.Dense(VOCAB_SIZE)(x)
+  model = keras.Model(inputs=inputs, outputs=outputs)
+  loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+  model.compile(
+      "adam", loss=loss_fn,
```
Comment (fchollet) [inline examples/nlp/text_generation_transformer_decoder.py:145]: Add metrics and use a keyword argument for the optimizer

File: examples/nlp/text_generation_transformer_decoder.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+      else:
+        x = start_tokens
+      x = np.array([x])
+
+      y = self.model.predict(x)
+      sample_token = self.sample_from(y[0][sample_index])
+      start_tokens.append(sample_token)
+      num_tokens_generated += 1
+    txt = self.tokenizer.detokenize(start_tokens)
+    print(f"generated text: \n{txt}\n")
```
[the remark this answers is not in the document]
Comment (fchollet) [reply examples/nlp/text_generation_transformer_decoder.py:212]: +1

File: examples/nlp/text_generation_transformer_decoder.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+start_prompt = "the"
+# Unpadded token sequence
+start_tokens = [tokenizer.token_to_id(token) for token in start_prompt.split()]
+text_gen_callback = TopKTextGenerator(
+    NUM_TOKENS_GENERATED, start_tokens, tokenizer
+)
+
+model = create_model()
+
+model.fit(
```
Comment (fchollet) [inline examples/nlp/text_generation_transformer_decoder.py:223]: Add a section for evaluation / inference and a conclusion section summarizing what was learned from the example.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: examples/nlp/text_generation_transformer_decoder.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
+"""
+Title: Text Generation with Keras NLP TransformerDecoder
+Author: Jesse Chan
+Date created: 2022/06/13
+Last modified: 2022/06/13
+Description: Implementation of a small GPT-like model using the TransformerDecoder class.
+"""
+"""
+# Download Library
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply examples/nlp/text_generation_transformer_decoder.py]: edited to ##

File: examples/nlp/text_generation_transformer_decoder.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
+"""
+Title: Text Generation with Keras NLP TransformerDecoder
+Author: Jesse Chan
+Date created: 2022/06/13
+Last modified: 2022/06/13
+Description: Implementation of a small GPT-like model using the TransformerDecoder class.
+"""
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply examples/nlp/text_generation_transformer_decoder.py]: added introduction section, with a high level description of the components in the notebook

File: examples/nlp/text_generation_transformer_decoder.py (hunk 1 above)
Comment (jessechancy) [reply examples/nlp/text_generation_transformer_decoder.py]: fixed

File: examples/nlp/text_generation_transformer_decoder.py (hunk 3 above)
Comment (jessechancy) [reply examples/nlp/text_generation_transformer_decoder.py]: added inference section and concluding paragraph

File: examples/nlp/text_generation_gpt.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+
+We strip the first 1000 vocabulary that are all of the form `"[unusedXX]"`, since they
+are all not useful for this text corpus and unneccesarily increases vocabulary size. Then
+we take the `VOCAB_SIZE` most frequent vocabularies after that. There are also some
+reserved tokens, so we want to make sure that `"[PAD]"` is at index 0 and `"[UNK]"` is in
+the vocabulary, since it is used for padding and unknown words respectively.
+"""
+
+# Download vocabulary data.
+vocab_file = keras.utils.get_file(
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:101]: should we use the word piece vocab learner utility here?

File: examples/nlp/text_generation_gpt.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+- One `keras_nlp.layers.TokenAndPositionEmbedding` layer, which combines the embedding
+for the token and its position. 
+- Multiple `keras_nlp.layers.TransformerDecoder` layers, with the default causal masking.
+The layer has no cross-attention when run with decoder sequence only.
+- One output dense linear layer
+"""
+
+
+def create_model():
+    inputs = keras.layers.Input(shape=(SEQ_LEN,), dtype=tf.int32)
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:183]: given that this is only called once, why not just move this out of the function?

File: examples/nlp/text_generation_gpt.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+    perplexity = keras_nlp.metrics.Perplexity(from_logits=True, mask_token_id=0)
+    model.compile(optimizer="adam", loss=loss_fn, metrics=[perplexity])
+    return model
+
+
+model = create_model()
+
+"""
+Let's take a look at our model summary here! We can see that a large majority of the
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:211]: remove the exclamation point here, so the one in the next sentence hits better.

File: examples/nlp/text_generation_gpt.py [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+SEQ_LEN = 128
+MIN_TRAINING_SEQ_LEN = 450
+
+# Inference
+NUM_TOKENS_TO_GENERATE = 80
+
+# Training
+LEARNING_RATE = 5e-4
+EPOCHS = 12
+NUM_TRAINING_BATCHES = 1000
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:70]: why do we need this hyperparameter, can we just train over the full dataset?

File: examples/nlp/text_generation_gpt.py [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+
+"""
+## Settings / Hyperparameters
+"""
+
+# Data
+SEQ_LEN = 128
+MIN_TRAINING_SEQ_LEN = 450
+
+# Inference
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:64]: maybe reorder the Hyperparameters blocks to

data
model
training
inference

File: examples/nlp/text_generation_gpt.py [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+# Load simplebooks-92 train set and filter out short lines.
+train_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/train.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+    .shuffle(buffer_size=256)
+)
+
+# Load simplebooks-92 validation set and filter out short lines.
+valid_ds = (
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:141]: valid -> val (here and elsewhere)

File: examples/nlp/text_generation_gpt.py [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+
+def preprocess(inputs):
+    outputs = tokenizer(inputs)
+    features = outputs[:, :-1]
+    labels = outputs[:, 1:]
+    return features, labels
+
+
+# Tokenize and split into train and label sequences.
+ds = train_ds.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:162]: let's keep the train in the name here of the dataset here... maybe raw_train_ds/train_ds and raw_val_ds/val_ds ?

File: examples/nlp/text_generation_gpt.py [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+"""
+## Inference
+
+With our trained model, we can test it out to gauge it's performance. Since
+this is a dataset of mostly fictional books, there is bound to be a hero, so let's use
+"The hero" as our starting string! We run it through the tokenizer to get the input for
+the model. 
+"""
+
+MAX_PREDICT_LEN = 80
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:238]: I think you don't use this

File: examples/nlp/text_generation_gpt.py [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+"""
+
+model.fit(ds.take(NUM_TRAINING_BATCHES), validation_data=ds_valid, verbose=2, epochs=EPOCHS)
+
+"""
+## Inference
+
+With our trained model, we can test it out to gauge it's performance. Since
+this is a dataset of mostly fictional books, there is bound to be a hero, so let's use
+"The hero" as our starting string! We run it through the tokenizer to get the input for
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:234]: remove exclamation point

File: examples/nlp/text_generation_gpt.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+"""
+## Tokenize Data
+
+We pre-process the dataset by tokenizing and splitting it into `features` and `labels`.
+"""
+
+
+def preprocess(inputs):
+    outputs = tokenizer(inputs)
+    features = outputs[:, :-1]
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:156]: should we add bos tokens here? then you could sample without seed text, and I think we would be a little closer to actual gpt

File: examples/nlp/text_generation_gpt.py [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+With our trained model, we can test it out to gauge it's performance. Since
+this is a dataset of mostly fictional books, there is bound to be a hero, so let's use
+"The hero" as our starting string! We run it through the tokenizer to get the input for
+the model. 
+"""
+
+MAX_PREDICT_LEN = 80
+start_prompt = "The hero"
+# Unpadded token sequence.
+start_tokens = [tokenizer.token_to_id(_) for _ in start_prompt.lower().split()]
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:241]: this feels awkward, we should actually use a tokenizer to tokenize the text. maybe let's just instantiate the tokenizer without a sequence length, and either use ragged.to_dense() or a packer layer to densify (especially if we decide to add start tokens)

File: examples/nlp/text_generation_gpt.py [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+        txt = tokenizer.detokenize(output_tokens)
+        print(f"Top-K search generated text: \n{txt}\n")
+
+
+text_generation_callback = TopKTextGenerator(k=10)
+# Dummy training loop to demonstrate callback.
+model.fit(ds.take(1), verbose=2, epochs=2, callbacks=[text_generation_callback])
+
+"""
+## Conclusion
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:411]: i'm not sure this conclusion section adds too much. we should either replace it with a further reading/where to next section, or remove it entirely

File: examples/nlp/text_generation_gpt.py [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+Pre-trained (GPT) model. GPT is a transformer based model that allows you to generate
+sophisticated text from a small input.
+
+We will train the model on the [simplebooks-92](https://arxiv.org/abs/1911.12391) corpus,
+which is a dataset made from several novels. It is a good dataset for this example since
+it has a small vocabulary and high word frequency, which is beneficial when training a
+model with few parameters.
+
+This example combines concepts from [Text generation with a miniature GPT](https://keras.io/examples/generative/text_generation_with_miniature_gpt/) 
+with KerasNLP abstractions. We will demonstrate how KerasNLP tokenization, model, metrics, and
```
Comment (mattdangerw) [inline examples/nlp/text_generation_gpt.py:22]: Maybe just:
We will demonstrate how KerasNLP tokenization, layers and metrics simplify the training
process, and then show how to generate output text using sampling utilities.

And then remove the whole next paragraph. Readers can read on to see the exact layers used.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
This is really great! Left a few comments

File: examples/nlp/text_generation_gpt.py (hunk 14 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:234]: edited

File: examples/nlp/text_generation_gpt.py (hunk 10 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:64]: edited

File: examples/nlp/text_generation_gpt.py (hunk 11 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:141]: edited

File: examples/nlp/text_generation_gpt.py (hunk 13 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:238]: edited

File: examples/nlp/text_generation_gpt.py (hunk 12 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:162]: edited

File: examples/nlp/text_generation_gpt.py (hunk 9 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:70]: removed hyperparameter

File: examples/nlp/text_generation_gpt.py [hunk 19]
```diff
@@ -0,0 +1,418 @@
...
+while the number of transformer decoder layers (`NUM_LAYERS`) doesn't affect it much.
+"""
+
+model.summary()
+
+"""
+## Training
+
+Now that we have our model, let's train it. We use a subset of the training data to save
+on training time. It would also be beneficial to use a GPU to speed up the training
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:223]: edited

File: examples/nlp/text_generation_gpt.py (hunk 7 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:183]: edited

File: examples/nlp/text_generation_gpt.py [hunk 20]
```diff
@@ -0,0 +1,418 @@
...
+    embedding_layer = keras_nlp.layers.TokenAndPositionEmbedding(
+        vocabulary_size=VOCAB_SIZE,
+        sequence_length=SEQ_LEN,
+        embedding_dim=EMBED_DIM,
+        mask_zero=True,
+    )
+    x = embedding_layer(inputs)
+    # Transformer decoders.
+    for _ in range(NUM_LAYERS):
+        transformer_block = keras_nlp.layers.TransformerDecoder(
```
[the remark this answers is not in the document]
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:194]: edited

File: examples/nlp/text_generation_gpt.py (hunk 6 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:101]: edited

File: examples/nlp/text_generation_gpt.py (hunk 18 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:22]: edited

File: examples/nlp/text_generation_gpt.py (hunk 8 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:211]: edited

File: examples/nlp/text_generation_gpt.py (hunk 15 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:156]: edited

File: examples/nlp/text_generation_gpt.py (hunk 16 above)
Comment (jessechancy) [reply examples/nlp/text_generation_gpt.py:241]: edited to use packer layer for start tokens

File: examples/nlp/text_generation_gpt.py [hunk 21]
```diff
@@ -0,0 +1,401 @@
+"""
+Title: Simple GPT Text Generation with KerasNLP transformers
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:2]: Just "with KerasNLP"  (no transformers)

File: examples/nlp/text_generation_gpt.py [hunk 22]
```diff
@@ -0,0 +1,401 @@
...
+Author: [Jesse Chan](https://github.com/jessechancy)
+Date created: 2022/07/25
+Last modified: 2022/07/25
+Description: Using KerasNLP transformers to train a mini-GPT model for text generation.
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:6]: Remove "transformers"

File: examples/nlp/text_generation_gpt.py [hunk 23]
```diff
@@ -0,0 +1,401 @@
...
+"""
+
+"""
+## Introduction
+
+In this example, we will use KerasNLP layers to build a scaled down Generative
+Pre-trained (GPT) model. GPT is a transformer based model that allows you to generate
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:13]: - Capitalize T

Comment (fchollet) [reply examples/nlp/text_generation_gpt.py:13]: Also capitalize Transformer

File: examples/nlp/text_generation_gpt.py [hunk 24]
```diff
@@ -0,0 +1,401 @@
...
+sophisticated text from a small input.
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:14]: "from a prompt"

File: examples/nlp/text_generation_gpt.py [hunk 25]
```diff
@@ -0,0 +1,401 @@
...
+which is a dataset made from several novels. It is a good dataset for this example since
+it has a small vocabulary and high word frequency, which is beneficial when training a
+model with few parameters.
+
+This example combines concepts from [Text generation with a miniature GPT](https://keras.io/examples/generative/text_generation_with_miniature_gpt/)
+with KerasNLP abstractions. We will demonstrate how KerasNLP tokenization, layers and
+metrics simplify the training
+process, and then show how to generate output text using sampling utilities.
+
+Note: If you are running this on a colab make sure to enable GPU runtime for faster
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:26]: "on Colab"

File: examples/nlp/text_generation_gpt.py [hunk 26]
```diff
@@ -0,0 +1,401 @@
...
+training performance.
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:27]: "faster training" (drop "performance")

File: examples/nlp/text_generation_gpt.py [hunk 27]
```diff
@@ -0,0 +1,401 @@
...
+"""
+
+"""
+# Mini-GPT Text Generation with KerasNLP transformers
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:31]: Drop "transformers"

Comment (fchollet) [reply examples/nlp/text_generation_gpt.py:31]: Also make sure that all section titles start with `##` and not `#`

File: examples/nlp/text_generation_gpt.py [hunk 28]
```diff
@@ -0,0 +1,401 @@
...
+LEARNING_RATE = 5e-4
+EPOCHS = 6
+
+# Inference
+NUM_TOKENS_TO_GENERATE = 80
+
+"""
+## Load Data
+
+Now, let's download the dataset! The SimpleBooks dataset has a small vocabulary, which
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:72]: Let's quantify this statement: what's the vocabulary size?

File: examples/nlp/text_generation_gpt.py [hunk 29]
```diff
@@ -0,0 +1,401 @@
...
+
+"""
+## Train Tokenizer
+
+We train the tokenizer from the training dataset for a vocabulary size of `VOCAB_SIZE`,
+which is a tuned hyperparameter. We want to limit the vocabulary as much as possible, as
+we will see later on
+that it has a large affect on the number of model parameters. We also don't want to take
+too few vocabulary, or there would be too many out of vocabulary (OOV) sub-words. In
+addition, three tokens are reserved in the vocabulary:
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:105]: Introduce line break before list

File: examples/nlp/text_generation_gpt.py [hunk 30]
```diff
@@ -0,0 +1,401 @@
...
+"""
+
+tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=vocab, sequence_length=SEQ_LEN
+)
+
+"""
+## Tokenize Data
+
+We pre-process the dataset by tokenizing and splitting it into `features` and `labels`.
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:138]: "preprocess" (single word)

File: examples/nlp/text_generation_gpt.py [hunk 31]
```diff
@@ -0,0 +1,401 @@
...
+
+With our trained model, we can test it out to gauge it's performance. Since this model is
+built with a `"[BOS]"` token, we can have an empty starting prompt for text generation.
+"""
+
+# Unpadded bos token.
+prompt_tokens = tf.convert_to_tensor([tokenizer.token_to_id("[BOS]")])
+
+"""
+We will use the `keras_nlp.utils` library for inference. Every text generation
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:227]: Say "module" rather than library

File: examples/nlp/text_generation_gpt.py [hunk 32]
```diff
@@ -0,0 +1,401 @@
...
+utility would require a `token_logits_fn()` wrapper around the model. This wrapper takes
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:228]: "requires"

File: examples/nlp/text_generation_gpt.py [hunk 33]
```diff
@@ -0,0 +1,401 @@
...
+
+# Load simplebooks-92 validation set and filter out short lines.
+raw_val_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/valid.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+)
+
+"""
+## Train Tokenizer
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:98]: In every section title, only capitalize the first word, not all words

File: examples/nlp/text_generation_gpt.py (hunk 21 above)
Comment (fchollet) [reply examples/nlp/text_generation_gpt.py:2]: Also don't capitalize all words unless they're proper nouns

File: examples/nlp/text_generation_gpt.py [hunk 34]
```diff
@@ -0,0 +1,401 @@
...
+
+
+text_generation_callback = TopKTextGenerator(k=10)
+# Dummy training loop to demonstrate callback.
+model.fit(train_ds.take(1), verbose=2, epochs=2, callbacks=[text_generation_callback])
+
+"""
+## Conclusion
+
+Congrats, you made it through the example! To recap, in this example, we use KerasNLP
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:393]: Drop "Congrats, you made it through the example!"

File: examples/nlp/text_generation_gpt.py [hunk 35]
```diff
@@ -0,0 +1,401 @@
...
+layers to train a sub-word vocabulary, tokenize training data, create a miniature GPT
+model, and perform inference with the text generation library.
+
+If you would like to understand how transformers work, or learn more about training the
+full GPT model, here are some further readings:
```
Comment (fchollet) [inline examples/nlp/text_generation_gpt.py:398]: Add line break before list

**Review by fchollet — COMMENTED** [review]
Thanks for the updates! I did a round of reviews with a focus on copywriting.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! I pushed some copyedits. Please pull them first. I think we're ready to add the generated files now.

**Review by fchollet — APPROVED** [review]
Thank you for the great contribution! 👍 Merging now.

**Review by fchollet — APPROVED** [review]

**Review by mattdangerw — APPROVED** [review]
LGTM! These examples are great! Really demonstrative of the different sampling.

Looks like we will need to cut a new release including the tokenizer vocab trainer function before we release this.

**Pull request merged**